### PR TITLE
PDCL-5968 - Response Tokens: return all propositions with response tokens

### PR DESCRIPTION
**Pull request by ninaceban**
## Description
Response tokens is a feature that is used by our customers mainly to send data to third parties, like google analytics, facebook etc. Every item within a decision object that will be returned from `konductor` will contain a `meta` block which contains the response tokens.
As we decided to proceed with a simple solution we will add `render:true` to decisions that we attempted to render. So that the customer could identify which decision items were attempted to render, so that he could then send the response tokens further. We will return  a `propositions` array and we will still return `decisions` array for backward compatibility.

note: Only the `propositions` array will contain decisions with `render:true` flag.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [ ] I have made any necessary test changes and all tests pass.
- [ ] I have run the Sandbox successfully.

### CODE REVIEW

File: src/components/Personalization/createAutoRenderingHandler.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import addRenderToExecutedDecisions from "./utils/addRenderToExecutedDecisions";
+
+export default ({
+  viewCache,
+  executeDecisions,
+  executeCachedViewDecisions,
+  showContainers
+}) => {
+  return ({ viewName, pageWideScopeDecisions, formBasedComposedDecisions }) => {
```
Comment (jfkhoury) [inline src/components/Personalization/createAutoRenderingHandler.js:9]: I recommend we rename those to something more generic that would apply to ODE, AJO...

`formBasedComposedDecisions` & `pageWideScopeDecisions`

Maybe something like: `renderableDecisions` & `nonRenderableDecisions`.

Also, aren't we renaming `decisions` to `propositions`?

Comment (ninaceban) [reply src/components/Personalization/createAutoRenderingHandler.js:9]: Thank you, Joe. I will rename those. 
Yes we return `decisions` the way we did earlier and newly `propositions` with decisions with rendered flag.

**Comment by Aaronius** [comment]
@ninaceban how do we test the non-rendering propositions in the sandbox? Are there any examples of that set up?

**Comment by Aaronius** [comment]
Oh, I think that's the `decisionScopes: ["sandbox-personalization-page"]` you've commented out. Maybe we should request the scope on the personalization page or something like that and "manually" render the proposition to the UI in some form so we can see that it's working?

File: src/components/Personalization/decisionsExtractor.js [hunk 2]
```diff
@@ -75,11 +77,35 @@ const extractDecisionsByScope = (decisions, scope) => {
...
+};
+
 export default {
-  groupDecisionsBySchema({ decisions, schema }) {
-    return splitDecisions(decisions, schema);
-  },
-  groupDecisionsByScope({ decisions, scope }) {
-    return extractDecisionsByScope(decisions, scope);
+  groupDecisions(decisions) {
+    return getGroupedDecisions(decisions);
```
Comment (Aaronius) [inline src/components/Personalization/decisionsExtractor.js:109]: I really like this change. I think it cleaned up a lot of the code that depended on it. Consider exporting this function as the default and renaming the module accordingly.

File: test/unit/specs/components/Personalization/createNonRenderingHandler.spec.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
```
Comment (Aaronius) [inline test/unit/specs/components/Personalization/createNonRenderingHandler.spec.js:2]: 2021

File: test/unit/specs/components/Personalization/utils/addRenderToExecutedDecisions.spec.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      {
+        blah: "123"
+      },
+      {
+        blah: "345"
+      }
+    ];
+    const result = addRenderToExecutedDecisions(decisions);
+    expect(result[0].rendered).toEqual(true);
+    expect(result[1].rendered).toEqual(true);
```
Comment (Aaronius) [inline test/unit/specs/components/Personalization/utils/addRenderToExecutedDecisions.spec.js:27]: I think it's important to assert that `decisions[0].rendered` and `decisions[1].rendered` are still `undefined`.

**Review by Aaronius — COMMENTED** [review]
This is looking good, @ninaceban! Thanks!

A few remaining items:

* We need functional tests.

* In the JIRA issue, I wrote this regarding the `decisions` property on the response object:

> For backward-compatibility, this needs to stay the same (still returning only form-based decisions), but when the property is accessed, a warning needs to be logged to the console explaining that the decisions property has been deprecated and that the customer should start using a new propositions property.

I still think we should do this. That means creating a getter for the `decisions` property.

* Let's finalize the name for the property currently named `rendered`. I suggested `isRenderAttempted` (to be clear that we've attempted to render it but not promising that it will end up succeeding). Artur preferred `autoRenderingEnabled`, which I'm also fine with, though I'd shorten it to `autoRenderEnabled`. @jonsnyder what's your take?

File: src/components/Personalization/createOnResponseHandler.js [hunk 5]
```diff
@@ -9,87 +9,60 @@ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTA
...
-    if (viewName && nonPageWideScopeDecisions[viewName]) {
-      decisionsToBeReturned.push(...nonPageWideScopeDecisions[viewName]);
+      return autoRenderingHandler({
+        viewName,
+        pageWideScopeDecisions,
+        formBasedComposedDecisions
+      });
     }
-
-    return { decisions: decisionsToBeReturned };
```
Comment (Aaronius) [inline src/components/Personalization/createOnResponseHandler.js:93]: This file is cleaning up well.

File: src/components/Personalization/createNonRenderingHandler.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        ...pageWideScopeDecisions,
+        ...formBasedComposedDecisions
+      ],
+      propositions: [
+        ...redirectDecisions,
+        ...pageWideScopeDecisions,
+        ...formBasedComposedDecisions
+      ]
+    };
+  };
```
[the remark this answers is not in the document]
Comment (ninaceban) [reply src/components/Personalization/createNonRenderingHandler.js:39]: thank you Aaron!

File: test/unit/specs/components/Personalization/utils/addRenderToExecutedDecisions.spec.js (hunk 4 above)
Comment (ninaceban) [reply test/unit/specs/components/Personalization/utils/addRenderToExecutedDecisions.spec.js:27]: thank you

File: src/components/Personalization/createOnResponseHandler.js (hunk 5 above)
Comment (ninaceban) [reply src/components/Personalization/createOnResponseHandler.js:93]: thank you, yes, that was my intention

File: test/unit/specs/components/Personalization/createNonRenderingHandler.spec.js (hunk 3 above)
Comment (ninaceban) [reply test/unit/specs/components/Personalization/createNonRenderingHandler.spec.js:2]: thank you

File: src/components/Personalization/decisionsExtractor.js (hunk 2 above)
Comment (ninaceban) [reply src/components/Personalization/decisionsExtractor.js:109]: thank you :)

File: src/components/Personalization/createExecuteCachedViewDecisions.js [hunk 7]
```diff
@@ -12,17 +12,15 @@ governing permissions and limitations under the License.
...
-      if (isNonEmptyArray(viewDecisions)) {
-        executeViewDecisions(viewDecisions);
-        return;
-      }
-      const xdm = { web: { webPageDetails: { viewName } } };
+export default ({ executeViewDecisions, collect }) => {
+  return ({ viewName, viewDecisions }) => {
+    if (isNonEmptyArray(viewDecisions)) {
+      executeViewDecisions(viewDecisions);
+      return;
```
[the remark this answers is not in the document]
Comment (ninaceban) [reply src/components/Personalization/createExecuteCachedViewDecisions.js:19]: sure thing!

File: src/components/Personalization/createNonRenderingHandler.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+export default ({ viewCache }) => {
```
[the remark this answers is not in the document]
Comment (ninaceban) [reply src/components/Personalization/createNonRenderingHandler.js:1]: thank you

File: src/components/Personalization/createAutoRenderingHandler.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        executeDecisions(pageWideScopeDecisions);
+        executeCachedViewDecisions({
+          viewName,
+          viewDecisions: currentViewDecisions
+        });
+        showContainers();
+
+        return {
+          get decisions() {
+            logger.warn("Decisions property is deprecated");
```
Comment (jfkhoury) [inline src/components/Personalization/createAutoRenderingHandler.js:34]: Awesome. Should we add: `. Please use "propositions" instead.` ?

File: src/components/Personalization/createAutoRenderingHandler.js (hunk 1 above)
Comment (jfkhoury) [reply src/components/Personalization/createAutoRenderingHandler.js:22]: @ninaceban are we going to rename those props?

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    expect(executeDecisions).toHaveBeenCalledWith(pageWideScopeDecisions);
+    expect(executeCachedViewDecisions).toHaveBeenCalledWith({
+      viewName,
+      viewDecisions: CART_VIEW_DECISIONS
+    });
+    expect(showContainers).toHaveBeenCalled();
+
+    result.decisions.forEach(decision => {
+      if (decision.scope === "__view__" || decision.scope === viewName) {
+        expect(decision.rendered).toEqual(true);
```
Comment (jfkhoury) [inline test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:71]: @ninaceban I agree with @Aaronius , I think we should change `rendered` to something more generic, like `hasAttemptedRendering` or whatever he was suggesting :)

File: src/components/Personalization/createAutoRenderingHandler.js (hunk 9 above)
Comment (Aaronius) [reply src/components/Personalization/createAutoRenderingHandler.js:34]: I like that idea. Let's use complete sentences as well, please. Something like `The decisions property is deprecated. Please use the propositions property instead.`

File: src/components/Personalization/createExecuteCachedViewDecisions.js [hunk 11]
```diff
@@ -14,10 +14,12 @@ import isNonEmptyArray from "../../utils/isNonEmptyArray";
 
 export default ({ executeViewDecisions, collect }) => {
   return ({ viewName, viewDecisions }) => {
+    // if there are viewDecisions for current view we will execute them and then send the collect call
     if (isNonEmptyArray(viewDecisions)) {
       executeViewDecisions(viewDecisions);
-      return;
+      return; // return here is to avoid the following code to be executed, that one is meant for the condition when viewDecisions is empty
```
Comment (Aaronius) [inline src/components/Personalization/createExecuteCachedViewDecisions.js:20]: Hey @ninaceban, I was hoping for more of the _why_ and less of the _what_. I think the _why_ should answer the question, "_Why_ don't we want to send an event when there are view decisions?" For example, is it because we're sending events inside of `executeViewDecisions` in that case so we don't need to send an event here?

File: src/components/Personalization/createNonRenderingHandler.js [hunk 12]
```diff
@@ -1,40 +1,54 @@
...
-        ...redirectDecisions,
-        ...pageWideScopeDecisions,
-        ...formBasedComposedDecisions
-      ]
-    };
+    return Promise.resolve(propositions)
+      .then(items =>
+        getViewPropositions({ viewCache, viewName, propositions: items })
+      )
+      .then(items => buildFinalResult({ logger, propositions: items }));
```
Comment (Aaronius) [inline src/components/Personalization/createNonRenderingHandler.js:52]: Can we shorten this to:

```
return getViewPropositions({ viewCache, viewName, propositions })
  .then(items => buildFinalResult({ logger, propositions: items }));
```

File: src/components/Personalization/createAutoRenderingHandler.js (hunk 9 above)
Comment (Aaronius) [reply src/components/Personalization/createAutoRenderingHandler.js:34]: Also, let's de-dupe this string into a shared constant as well, since it's used in a three different places and doesn't minify.

File: src/components/Personalization/groupDecisions.js [hunk 13]
```diff
@@ -104,8 +104,6 @@ const getGroupedDecisions = unprocessedDecisions => {
   };
 };
 
-export default {
-  groupDecisions(decisions) {
-    return getGroupedDecisions(decisions);
-  }
+export default decisions => {
+  return getGroupedDecisions(decisions);
 };
```
Comment (Aaronius) [inline src/components/Personalization/groupDecisions.js:109]: Preferably rename `getGroupedDecisions` to `groupDecisions` and then just shorten this to `export default groupDecisions;`.

File: src/components/Personalization/utils/addRenderToExecutedDecisions.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { assign } from "../../../utils";
+
+export default decisions => {
+  return decisions.map(decision => assign({ rendered: true }, decision));
+};
```
Comment (Aaronius) [inline src/components/Personalization/utils/addRenderToExecutedDecisions.js:17]: This module should probably be named consistently with whatever the property ends up being named. Right now the property is named `rendered`, which means the module should probably be named `addRenderedToExecutedDecisions` rather than `addRenderToExecutedDecisions`.

**Comment by Aaronius** [comment]
@ninaceban, for consistency, I think the `rendered` property (or whatever it ends up being named) should be on every proposition returned to the customer. It will have a value of `true` for some propositions and `false` for others. Currently, the property only exists on the proposition if the value is `true`.

**Comment by ninaceban** [comment]
> @ninaceban, for consistency, I think the `rendered` property (or whatever it ends up being named) should be on every proposition returned to the customer. It will have a value of `true` for some propositions and `false` for others. Currently, the property only exists on the proposition if the value is `true`.

@Aaronius  I was under the impression that we should add the `rendered` flag only when `renderDecisions` is set to `true`. If PMs think that we should set this flag all the time to all type of decisions we can address it later.

File: src/components/Personalization/utils/addRenderToExecutedDecisions.js (hunk 14 above)
Comment (ninaceban) [reply src/components/Personalization/utils/addRenderToExecutedDecisions.js:17]: Agree, but let's decide on the name first!

File: src/components/Personalization/groupDecisions.js (hunk 13 above)
Comment (ninaceban) [reply src/components/Personalization/groupDecisions.js:109]: ok, I will rename it. Thanks

File: src/components/Personalization/createExecuteCachedViewDecisions.js (hunk 11 above)
Comment (ninaceban) [reply src/components/Personalization/createExecuteCachedViewDecisions.js:20]: Ok, I will change the comment

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js (hunk 10 above)
Comment (ninaceban) [reply test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:71]: ok, let's decide during our checkin meeting.

File: src/components/Personalization/createAutoRenderingHandler.js (hunk 9 above)
Comment (ninaceban) [reply src/components/Personalization/createAutoRenderingHandler.js:34]: ok, will address the changes

**Comment by Aaronius** [comment]
After discussion amongst the team during Tech check-in, we've made some additional decisions/clarifications:

- The boolean property on each proposition that indicates whether the web SDK has attempted to render the proposition should be renamed from rendered to renderAttempted. This more appropriately reflects that the web SDK has attempted to render the proposition, but not that it's necessarily complete or whether it will succeed or fail.
- For propositions within the propositions array returned from a sendEvent command, each proposition will always have a renderAttempted property. Sometimes the value may be true and other times it will be false, but it will always exist. This provides a more predictable API since the property can always be relied on existing.

File: src/components/Personalization/decisionsExtractor.js [hunk 15]
```diff
@@ -75,11 +77,35 @@ const extractDecisionsByScope = (decisions, scope) => {
...
   }
```
Comment (jonsnyder) [inline src/components/Personalization/decisionsExtractor.js]: Since there is only one function that you are returning, you could change this filename to groupDecisions.js, and just export the one function.

File: src/components/Personalization/createAutoRenderingHandler.js (hunk 1 above)
Comment (jonsnyder) [reply src/components/Personalization/createAutoRenderingHandler.js:22]: I agree. I now know that "formBased" => non auto renderable and "domAction" => auto renderable. I think it makes things more clear to use renderable and nonRenderable.

**Review by jonsnyder — APPROVED** [review]

File: src/components/Personalization/decisionsExtractor.js (hunk 15 above)
Comment (ninaceban) [reply src/components/Personalization/decisionsExtractor.js]: yes, it was already renamed to groupDecisions.

File: src/components/Personalization/constants/loggerMessages.js [hunk 16]
```diff
@@ -0,0 +1,5 @@
+export const decisionsDeprecatedWarning =
+  "The decisions property is decisionsDeprecatedWarning. Please use the propositions property instead.";
```
Comment (Aaronius) [inline src/components/Personalization/constants/loggerMessages.js:2]: I don't think you ended to have `decisionsDeprecatedWarning` in this text.

File: src/components/Personalization/constants/loggerMessages.js [hunk 17]
```diff
@@ -0,0 +1,5 @@
...
+export const authoringEnabled = "Rendering is disabled, authoring mode.";
```
Comment (Aaronius) [inline src/components/Personalization/constants/loggerMessages.js:3]: Maybe `Rendering is disabled for authoring mode.` instead.

File: src/components/Personalization/constants/loggerMessages.js [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+export const redirectExecutionError =
+  "An error occurred while executing the redirect offer.";
```
Comment (Aaronius) [inline src/components/Personalization/constants/loggerMessages.js:5]: For consistency, follow [these conventions](https://github.com/adobe/alloy/wiki/Conventions#enumerations), specifically using all caps for the constant names and changing the filename to `loggerMessage.js`

**Review by Aaronius — APPROVED** [review]
